portal-bridge/src/types/state/trie_walker.rs

```diff
@@ -0,0 +1,155 @@
+use alloy_primitives::{Bytes, B256};
+use eth_trie::{decode_node, node::Node};
+use hashbrown::HashMap as BrownHashMap;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct AccountProof {
+    pub path: Vec<u8>,
+    pub proof: Vec<Bytes>,
+}
+
+pub struct PathToParent {
+    pub parent_hash: B256,
+    pub path_nibbles: Vec<u8>,
+}
+
+impl PathToParent {
+    pub fn new(parent_hash: B256, path_nibbles: Vec<u8>) -> Self {
+        Self {
+            parent_hash,
+            path_nibbles,
+        }
+    }
+}
+
+/// This struct takes in a root hash and a hashmap of changed nodes, then you can call an iterator
+/// which will return every proof to gossip
+pub struct TrieWalker {
+    pub nodes: BrownHashMap<B256, Vec<u8>>,
+    paths_to_parent: BrownHashMap<B256, PathToParent>,
+}
+
+impl TrieWalker {
+    pub fn new(root_hash: B256, nodes: BrownHashMap<B256, Vec<u8>>) -> Self {
+        let paths_to_parent = Self::get_paths_backwards(root_hash, &nodes);
+        Self {
+            nodes,
+            paths_to_parent,
+        }
+    }
+
+    fn get_paths_backwards(
+        root_hash: B256,
+        nodes: &BrownHashMap<B256, Vec<u8>>,
+    ) -> BrownHashMap<B256, PathToParent> {
+        let mut paths_to_parent: BrownHashMap<B256, PathToParent> = BrownHashMap::new();
+        let mut stack = vec![root_hash];
+        while let Some(node_key) = stack.pop() {
+            let encoded_node = nodes
+                .get(&node_key)
+                .expect("The stack should only contain nodes that are in the changed nodes");
+
+            let decoded_node = decode_node(&mut encoded_node.as_slice())
+                .expect("Should should only be passing valid encoded nodes");
+
+            match decoded_node {
+                Node::Extension(extension) => {
+                    let extension = extension.read().expect("Reading an extension should work");
+                    if let Node::Hash(hash_node) = &extension.node {
+                        // We could be walking a partial trie, so we need to check if the node is in
```

Comment:
nit: While I understand what you wanted to say, I find this comment a bit confusing and I imagine somebody not fully familiar with functionality might as well.

I would add something like:
_Only process provided nodes (they belong to the partial trie that we care about)_